samples/guide/src/main/java/com/squareup/okhttp/recipes/UploadProgress.java

```diff
@@ -0,0 +1,120 @@
+
+import java.io.File;
+import java.io.IOException;
+
+public final class UploadProgress {
+  /**
+   * The imgur client ID for OkHttp recipes. If you're using imgur for anything
+   * other than running these examples, please request your own client ID!
+   * https://api.imgur.com/oauth2
+   */
+  private static final String IMGUR_CLIENT_ID = "9199fdef135c122";
+  private static final MediaType MEDIA_TYPE_PNG = MediaType.parse("image/png");
+
+  private final OkHttpClient client = new OkHttpClient();
+
+  public void run() throws Exception {
+    final ProgressListener progressListener = new ProgressListener() {
+      @Override public void update(long bytesWritten, long contentLength, boolean done) {
+        System.out.println(bytesWritten);
+        System.out.println(contentLength);
+        System.out.println(done);
+        System.out.format("%d%% done\n", (100 * bytesWritten) / contentLength);
+      }
+    };
+
+    // Use the imgur image upload API as documented at https://api.imgur.com/endpoints/image
+    Request request = new Request.Builder()
+        .header("Authorization", "Client-ID " + IMGUR_CLIENT_ID)
+        .url("https://api.imgur.com/3/image")
+        .post(new ProgressRequestBody(
+            MEDIA_TYPE_PNG,
+            new File("website/static/logo-square.png"),
+            progressListener))
+        .build();
+
+    Response response = client.newCall(request).execute();
+    if (!response.isSuccessful()) throw new IOException("Unexpected code " + response);
+
+    System.out.println(response.body().string());
+  }
+
+  public static void main(String... args) throws Exception {
+    new UploadProgress().run();
+  }
+
+  private static class ProgressRequestBody extends RequestBody {
+
+    private final ProgressListener progressListener;
+    private final MediaType contentType;
+    private final File file;
+
+    public ProgressRequestBody(MediaType contentType, File file,
+                               ProgressListener progressListener) {
+      this.contentType = contentType;
+      this.file = file;
+      this.progressListener = progressListener;
+    }
+
+    @Override public MediaType contentType() {
+      return contentType;
+    }
+
+    @Override public long contentLength() {
+      return file.length();
+    }
+
+    @Override public void writeTo(BufferedSink sink) throws IOException {
+      BufferedSink bufferedSink = Okio.buffer(sink(sink));
+      Source source = Okio.source(file);
+      bufferedSink.writeAll(source);
+      source.close();
+      bufferedSink.close();
+    }
+
+    public Sink sink(Sink sink) {
+      return new ForwardingSink(sink) {
+        long totalBytesWritten = 0L;
+
+        @Override public void write(Buffer source, long byteCount) throws IOException {
+          super.write(source, byteCount);
+          totalBytesWritten += byteCount;
+          progressListener.update(totalBytesWritten, contentLength(),
+              totalBytesWritten == contentLength());
```

Comment:
(You can look at the PostStreaming example to see an unknown content length in practice.)